quaterion/loss/cos_face_loss.py

```diff
@@ -0,0 +1,73 @@
+from typing import Optional
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch import LongTensor, Tensor
+
+from quaterion.loss.group_loss import GroupLoss
+from quaterion.utils import l2_norm
+
+
+class CosFaceLoss(GroupLoss):
+
+    """Large Margin Cosine Loss as defined in https://arxiv.org/pdf/1801.09414.pdf
+
+    Args:
+        embedding_size: Output dimension of the encoder.
+        num_groups: Number of groups in the dataset.
+        scale: Scaling value to make cross entropy work.
+        margin: Margin value to push groups apart.
+    """
+
+    def __init__(
+        self,
+        embedding_size: int,
+        num_groups: int,
+        margin: Optional[float] = 0.35,
+        scale: Optional[float] = 64.0,
+    ):
+
+        super(GroupLoss, self).__init__()
+
+        self.kernel = nn.Parameter(torch.FloatTensor(embedding_size, num_groups))
+        nn.init.normal_(self.kernel, std=0.01)
+        self.scale = scale
+        self.margin = margin
+
+    def forward(self, embeddings: Tensor, groups: LongTensor) -> Tensor:
+
+        """Compute loss value
+        Args:
+            embeddings: shape: (batch_size, vector_length) - Output embeddings from the
+                encoder.
+            groups: shape: (batch_size,) - Group ids associated with embeddings.
+        Returns:
+            Tensor: loss value.
+        """
+
```

Comment:
Let's make a proper assertion.

```suggestion
        assert groups.ge(0).all() and groups.lt(self.kernel.size(1)).all(), \
            f"Invalid group ids: all the values must be between 0 (inclusive) and num_groups (exclusive), but given:  {groups}"

```